File: calculator.py
from math import pi
from typing import List
from enum import Enum


class Orientation(Enum):
    SERIES = 0
    SHUNT = 1


class Element:
    def __init__(self, orientation: Orientation):
        self.orientation = orientation

        self.z_func = lambda _: 1e-14  # function to define series impedance
        self.y_func = lambda _: 1e14  # function to define admittance

    def z(self, freq):
        return self.z_func(freq)

    def y(self, freq):
        return self.y_func(freq)


class Capacitor(Element):
    def __init__(self, val, orientation: Orientation = Orientation.SERIES, q=None):
        Element.__init__(self, orientation)
        self.val = val
        if q is None:
            # ideal capacitor
            self.y_func = lambda freq: 1j * 2 * pi * freq * self.val
            self.z_func = lambda freq: 1.0 / self.y_func(freq)
        else:
            self.q = q
            self.y_func = lambda freq: (2 * pi * freq * self.val) / self.q + 1j * 2 * pi * freq * self.val
            self.z_func = lambda freq: 1.0j / self.z_func(freq)


class Inductor(Element):
    def __init__(self, val, orientation: Orientation = Orientation.SERIES, q=None):
        Element.__init__(self, orientation)
        self.val = val
        if q is None:
            # ideal inductor
            self.z_func = lambda freq: 1j * 2 * pi * freq * self.val
            self.y_func = lambda freq: 1.0 / self.z_func(freq)
        else:
            self.q = q
            self.z_func = lambda freq: (2 * pi * freq * self.val) / self.q + 1j * 2 * pi * freq * self.val
            self.y_func = lambda freq: 1.0j / self.z_func(freq)


class Network:
    def __init__(self, zl):
        self.elements: List[Element] = []
        self.zl = lambda _: zl  # Load impedance

    def compute_impedances(self, freq):
        z_arr = [self.zl(freq)]

        for element in self.elements:
            if element.orientation == Orientation.SERIES:
                z_arr.append(z_arr[-1] + element.z(freq))
            elif element.orientation == Orientation.SHUNT:
                z_arr.append(1.0 / (1.0 / z_arr[-1] + element.y(freq)))

        return z_arr

    def add_element(self, element: Element):
        self.elements.append(element)


if __name__ == "__main__":
    import numpy as np
    import matplotlib.pyplot as plt

    net = Network(9 + 87j)

    c1 = Capacitor(4.89e-12, Orientation.SHUNT)
    l1 = Inductor(14.67e-9, Orientation.SERIES)
    c2 = Capacitor(11.001e-12, Orientation.SHUNT)

    net.add_element(c1)
    net.add_element(l1)
    net.add_element(c2)

    print(net.compute_impedances(868e6))

    freqs = np.linspace(600e6, 1500e6, 1000)
    zin_list = [net.compute_impedances(f)[-1] for f in freqs]
    z0 = 50
    s11 = [(z - z0) / (z + z0) for z in zin_list]
    s11_db = 20 * np.log10(np.abs(s11))

    plt.plot(freqs / 1e6, s11_db)
    plt.xlabel("Frequency (MHz)")
    plt.ylabel("S11 (dB)")
    plt.grid(True)
    plt.show()

File: calculator.js
import Complex from './complex.js';

export function capacitorY(freq, c) {
  const w = 2 * Math.PI * freq;
  return new Complex(0, w * c);
}

export function capacitorZ(freq, c) {
  return Complex.div(1, capacitorY(freq, c));
}

export function inductorY(freq, l) {
  return Complex.div(1, inductorZ(freq, l));
}

export function inductorZ(freq, l) {
  const w = 2 * Math.PI * freq;
  return new Complex(0, w * l);
}

export function computeImpedances(network, zl, freq, directions = []) {
  const zArr = [zl];
  network.forEach((elem) => {
    if (elem.orientation === 'series') {
      if (elem.type === 'capacitor') {
        zArr.push(
          Complex.add(zArr[zArr.length - 1], capacitorZ(freq, elem.value))
        );
        directions.push("R'");
      } else if (elem.type === 'inductor') {
        zArr.push(
          Complex.add(zArr[zArr.length - 1], inductorZ(freq, elem.value))
        );
        directions.push('R');
      }
    } else {
      if (elem.type === 'capacitor') {
        zArr.push(
          Complex.div(
            1,
            Complex.add(
              Complex.div(1, zArr[zArr.length - 1]),
              capacitorY(freq, elem.value)
            )
          )
        );
        directions.push('G');
      } else if (elem.type === 'inductor') {
        zArr.push(
          Complex.div(
            1,
            Complex.add(
              Complex.div(1, zArr[zArr.length - 1]),
              inductorY(freq, elem.value)
            )
          )
        );
        directions.push("G'");
      }
    }
  });
  return zArr;
}

export function matchImpedance(zl, z0, f) {
  const z = Complex.div(zl, z0);
  const networks = [];

  if (z.re < 1) {
    let x1 = -z.im + Math.sqrt(z.re * (1 - z.re));
    let x2 = -z.im - Math.sqrt(z.re * (1 - z.re));

    let net = [];
    if (x1 < 0) {
      // series capacitor
      const c = 1 / (-x1 * z0) / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'series' });
    } else {
      // series inductor
      const l = (x1 * z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'series' });
    }
    let x3 = Complex.sub(
      new Complex(1, 0),
      Complex.inv(Complex.add(z, new Complex(0, x1)))
    ).im;
    if (x3 < 0) {
      // shunt inductor
      const l = 1 / (-x3 / z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'shunt' });
    } else {
      // shunt capacitor
      const c = x3 / z0 / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'shunt' });
    }
    networks.push(net);

    net = [];
    if (x2 < 0) {
      // series capacitor
      const c = 1 / (-x2 * z0) / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'series' });
    } else {
      // series inductor
      const l = (x2 * z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'series' });
    }
    x3 = Complex.sub(
      new Complex(1, 0),
      Complex.inv(Complex.add(z, new Complex(0, x2)))
    ).im;
    if (x3 < 0) {
      // shunt inductor
      const l = 1 / (-x3 / z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'shunt' });
    } else {
      // shunt capacitor
      const c = x3 / z0 / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'shunt' });
    }
    networks.push(net);
  }

  if (Complex.div(1, z).re < 1) {
    const denom = z.re * z.re + z.im * z.im;
    let x1 = z.im / denom + Math.sqrt((z.re / denom) * (1 - z.re / denom));
    let x2 = z.im / denom - Math.sqrt((z.re / denom) * (1 - z.re / denom));
    let net = [];

    if (x1 < 0) {
      // shunt inductor
      const l = 1 / (-x1 / z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'shunt' });
    } else {
      // shunt capacitor
      const c = x1 / z0 / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'shunt' });
    }
    let x3 = Complex.sub(
      new Complex(1, 0),
      Complex.inv(Complex.add(Complex.inv(z), new Complex(0, x1)))
    ).im;
    if (x3 < 0) {
      // series capacitor
      const c = 1 / (-x3 * z0) / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'series' });
    } else {
      // series inductor
      const l = (x3 * z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'series' });
    }
    networks.push(net);

    net = [];
    if (x2 < 0) {
      // shunt inductor
      const l = 1 / (-x2 / z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'shunt' });
    } else {
      // shunt capacitor
      const c = x2 / z0 / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'shunt' });
    }
    x3 = Complex.sub(
      new Complex(1, 0),
      Complex.inv(Complex.add(Complex.inv(z), new Complex(0, x2)))
    ).im;
    if (x3 < 0) {
      // series capacitor
      const c = 1 / (-x3 * z0) / (2 * Math.PI * f);
      net.push({ type: 'capacitor', value: c, orientation: 'series' });
    } else {
      // series inductor
      const l = (x3 * z0) / (2 * Math.PI * f);
      net.push({ type: 'inductor', value: l, orientation: 'series' });
    }
    networks.push(net);
  }

  return networks;
}
